python_modules/dagster/dagster/_core/executor/multiprocess.py

```diff
@@ -187,137 +188,143 @@ def execute(
-                stopping: bool = False
-
-                while (not stopping and not active_execution.is_complete) or active_iters:
-                    if active_execution.check_for_interrupts():
-                        yield DagsterEvent.engine_event(
-                            plan_context,
-                            (
-                                "Multiprocess executor: received termination signal - "
-                                "forwarding to active child processes"
-                            ),
-                            EngineEventData.interrupted(list(term_events.keys())),
-                        )
-                        stopping = True
-                        active_execution.mark_interrupted()
-                        for key, event in term_events.items():
-                            event.set()
-
-                    while not stopping:
-                        steps = active_execution.get_steps_to_execute(
-                            limit=(limit - len(active_iters)),
-                        )
-
-                        if not steps:
-                            break
-
-                        for step in steps:
-                            step_context = plan_context.for_step(step)
-                            term_events[step.key] = multiproc_ctx.Event()
-                            active_iters[step.key] = execute_step_out_of_process(
-                                multiproc_ctx,
-                                job,
-                                step_context,
-                                step,
-                                errors,
-                                term_events,
-                                self.retries,
-                                active_execution.get_known_state(),
-                                execution_plan.repository_load_data,
-                            )
-
-                    # process active iterators
-                    empty_iters = []
-                    for key, step_iter in active_iters.items():
-                        try:
-                            event_or_none = next(step_iter)
-                            if event_or_none is None:
-                                continue
-                            else:
-                                yield event_or_none
-                                active_execution.handle_event(event_or_none)
-
-                        except ChildProcessCrashException as crash:
-                            serializable_error = serializable_error_info_from_exc_info(
-                                sys.exc_info()
-                            )
-                            step_context = plan_context.for_step(
-                                active_execution.get_step_by_key(key)
-                            )
+            with GlobalConcurrencyContext(
+                plan_context.instance, plan_context.run_id
+            ) as global_concurrency_context:
+                with ActiveExecution(
```

Comment:
do these all exit at the same time? might be able to remove some indentation with an ExitStack